### ISSUE TITLE
[Bug] Using custom sync server does not work (server returns 404)

### ISSUE THREAD

**Issue by chrstn-hntschl** [mozilla-mobile/fenix#15590]
## Steps to reproduce
Deploy your own sync server (uising docker: mozilla/syncserver:latest) behind an nginx reverse proxy with ssl). Turn on sync on firefox desktop 81.0 works fine (setting identity.sync.tokenserver.uri = https://subdomain.domain.tld/token/1.0/sync/1.5 and using default firefox accounts server). Using the same settings in firefox android 81.1.1 (setting user-defined sync server to https://subdomain.domain.tld/token/1.0/sync/1.5) fails, firefox android shows "last sync: Jan, 1st, 1970". Debug message of self hosted syncserver shows: 
```
INFO:mozsvc.metrics:{"code": 404, "request_time": 0.002107858657836914, "remoteAddressChain": ["141.*.*.*", "172.*.*.*"], "agent": "MozacFetch/57.0.6", "path": "https://subdomain.domain.tld/1.0/sync/1.5", "method": "GET"}
```
Considering the log message, it seems as if firefox android is using the wrong server url - the 'token' part is stripped off. 

### Expected behavior
Firefox android should sync with self-hosted syncserver.

### Actual behavior
Sync fails.

### Device information

* Android device: Nokia 7.1, Android 10
* Fenix version: 81.1.1(Build #2015764547)

**Comment by tlambertz** [comment]
I can confirm this bug and found a workaround. I am using nginx as a reverseproxy in front of my syncserver and simply added a rule so that `url.tld/1.0/` proxies to `url.tld/token/1.0/`.

**Comment by chrstn-hntschl** [comment]
Thanks for the pointer @tlambertz  - I did the same, does work now! Yet this seems a true bug that should be resolved...

**Comment by ironashram** [comment]
i can confirm this behavior, it seems to me that if you had Firefox below 81 configured and then upgraded to latest ti still works fine, but fails on every new install

**Comment by tslpre** [comment]
Tested another workaround (which doesn't need to touch the web sever): just add a "/token" to the custom address - e.g.  https://server/token/token/1.0/sync/1.5

That works...

**Comment by SanSanysch** [comment]
> Tested another workaround (which doesn't need to touch the web sever): just add a "/token" to the custom address - e.g. https://server/token/token/1.0/sync/1.5
> 
> That works...

It bug!!! I suffer already 8 days! Write it to https://github.com/mozilla-services/syncserver

**Comment by pathwayx99** [comment]
I confirm the same issue. Been pulling my hair out to find out why I couldn't sync my new phone (OnePlus 8 running Android 10, Firefox for Android 82.1.1. 

I'm running Sync off of the latest official Docker container, properly configured to run behind Traefik with valid LetsEncrypt certs. Desktop clients syncing fine, my old OnePlus 5T running version 81.1.2 also syncing fine, my new phone zilch.

I side-loaded an older apk, used the exact same sync server URI and it worked perfectly, so it was nothing related to the phone and/or networking.

So I ran logcat via adb and isolated the issue to the 404 message as well. From my logs:

> 11-04 16:43:27.487  7111  8670 W sync15::token: Non-success status when fetching token: 404
11-04 16:43:27.487  7111  8670 W sync15::sync_multiple: sync failed: HTTP status 404 when requesting a token from the tokenserver, final status=ServiceError

I tried the work-around above (adding an extra "token") in the Nightly and it works perfectly. 

So I can also confirm this bug...

**Comment by rfk** [comment]
Thanks for the reports here folks! I think I've tracked this down to a bug in the underlying sync components, filed https://github.com/mozilla/application-services/issues/3693 for followup.

**Comment by rfk** [comment]
I've got a fix up for this over in https://github.com/mozilla/application-services/pull/3694, but it will take a little while for this to work its way through various releases and into the mobile browser.

In the meantime, I believe setting the URL to `https://example.com/token/` (with a trailing slash, but without the `1.0/sync/1.5` part) should work around the issue.

Using `https://example.com/token/ANYTHINGHERE/1.0/sync/1.5` as suggested in earlier comments will work *for now*, but it will stop working once that fix for this bug is released, because that's not a valid URL on the server. Since `https://example.com/token/` is a valid URL on the server it should keep working even once the fix is released.

**Comment by pathwayx99** [comment]
Thanks for that, Ryan. I'll test the proposed interim solution later today with the Nightly and will let you know how it worked out.

**Comment by pathwayx99** [comment]
Update: Using just "https://sub.domain.com/token/" in the Nightly does *not* work for syncing.

**Comment by rfk** [comment]
Oh, interesting! Thanks for trying it out @pathwayx99. So just to confirm my understanding:

* Using `https://sub.domain.com/token/token/1.0/sync/1.5` (with the extra `/token`) works for you.
* Using `https://sub.domain.com/token/` does not; does it give the same `HTTP status 404` error message as in comments above?

**Comment by pathwayx99** [comment]
Hmm. @rfk, I connected the phone with the computer this morning to run adb again and lo and behold it syncs. I tried it a couple of times and I verified that it got logins, bookmarks etc from the custom sync server with the `https://sub.domain.com/token/` URI.

So whatever happened last night might have been unrelated or transient.

Hence, I confirm that:

Using `https://sub.domain.com/token/token/1.0/sync/1.5` (with the extra /token) works for me.
Using `https://sub.domain.com/token/` also works for me.

**Comment by rfk** [comment]
Great, thanks for confirming!

**Comment by ericlathrop** [comment]
Hi, I'm having trouble getting FF mobile 82.1.3 to sync. I can login, but it always says "Last synced: never". Workarounds I've tried:
1. Setting custom sync server `https://sub.domain.com`
2. Setting custom sync server `https://sub.domain.com/token`
3. Setting custom sync server `https://sub.domain.com/token/`
4. Setting custom sync server `https://sub.domain.com/token/1.0/sync/1.5`
5. Setting custom sync server `https://sub.domain.com/token/token/1.0/sync/1.5`
6. Setting custom sync server `https://sub.domain.com/1.0/sync/1.5`
7. Server-side redirect of `/1.0/*` to `/token/1.0/$1`
8. Server-side rewrite of `/1.0/*` to `/token/1.0/$1`

Not sure what else to try.

**Comment by pathwayx99** [comment]
Hi @ericlathrop. Some things to verify (I'm looking at my own setup):

1. Have you set SYNCSERVER_PUBLIC_URL to the https:// address you use as sync server?
2. I have the following three environment variables set to true -- you? 

```
SYNCSERVER_BATCH_UPLOAD_ENABLED=true
SYNCSERVER_FORCE_WSGI_ENVIRON=true
SYNCSERVER_ALLOW_NEW_USER=true
```
3. In the traefik labels, I'm using my https entrypoint

`traefik.http.routers.sync.entrypoints=secure`

4. Are your letsencrypt certificates coming through ok? If you navigate to the sync server https:// via chrome, do you get a green mark?
5. Have you got your DNS pointing to the sync server address?

Hope something from the above sparks a thought or helps!

**Comment by ericlathrop** [comment]
@pathwayx99 I already have the syncserver working for the desktop browser, so I know it's setup correctly in some capacity.
1. yes
2. I've tried that combination, and a few others
3. same
4. yes
5. yes

**Comment by ericlathrop** [comment]
My problem turned out to be that I had a `http` URL in `SYNCSERVER_PUBLIC_URL` because I was terminating ssl on the reverse proxy (traefik) and didn't think the `https` URL would work. I had to set `SYNCSERVER_PUBLIC_URL` to the https URL, and additionally set `SYNCSERVER_FORCE_WSGI_ENVIRON` to `true`, then set my sync server URL in the Android app to `https://mydomain/token/` to make it work.

I don't understand why an incorrect `SYNCSERVER_PUBLIC_URL` breaks mobile, but not desktop, but at least I finally got it working.
